kernel/src/common/regs/README.md

```diff
@@ -0,0 +1,207 @@
+# Register/Bitfield Interface for Tock
+
+This module provides an interface for defining and manipulating memory mapped
+registers and bitfields.
+
+## Defining registers
+
+The module provides three types for working with memory mapped registers: `ReadWrite`,
+`ReadOnly`, and `WriteOnly`, providing read-write, read-only, and write-only functionality,
+respectively.
+
+Defining the registers is similar to the C-style approach, where each register
+is a field in a packed struct:
+
+```rust
+use common::regs::{ReadOnly, ReadWrite, WriteOnly};
+
+#[repr(C, packed)]
+struct Registers {
+    // Control register: read-write
+    // The 'Control' parameter constrains this register to only use fields from
+    // a certain group (defined below in the bitfields section).
+    cr: ReadWrite<u8, Control>,
+
+    // Status register: read-only
+    s: ReadOnly<u8, Status>
+
+    // Registers can be bytes, halfwords, or words:
+    // Note that the second type parameter can be omitted, meaning that there
+    // are no bitfields defined for these registers.
+    byte0: ReadWrite<u8>,
+    byte1: ReadWrite<u8>,
+    short: ReadWrite<u16>,
+    word: ReadWrite<u32>
+
+    // Etc.
+}
+```
+
+## Defining bitfields
+
+Bitfields are defined through the `bitfields!` macro:
+
+```rust
+bitfields! [
+    // First parameter is the register width for the bitfields. Can be u8, u16,
+    // or u32.
+    u8,
+
+    // Each subsequent parameter is a register abbreviation, its descriptive
+    // name, and its associated bitfields.
+    // The descriptive name defines this 'group' of bitfields. Only registers
+    // defined as ReadWrite<_, Control> can use these bitfields.
+    CR Control [
+        // Bitfields are defined as:
+        // name (Mask(mask), shift) [ /* optional named values */ ]
+
+        // This is a two-bit field which includes bits 4 and 5
+        RANGE (Mask(0b11), 4) [
```

Comment:
Can `Mask` ever be non-consecutive bits? Does it make more sense to just have a number, i.e. the size of the field? I will note I have no idea if this is possible.

Also, is there a way to not require a `(` here? It makes the syntax highlighting think that `RANGE` is a function (which is entirely plausible) making this structure difficult to parse (I spent a while trying to figure out what the `RANGE` and `MODE` functions did.

Reply:
With regard to `(`, it isn't required, it can be changed to `[` or `{`.

It would be nicer to specify the size of the field--the problem is that underneath the hood, I need the mask bits as a literal constant.

Reply:
I think a related question, what's the right way to handle "split" values (equivalently non-consecutive bits). A concrete example:

<img width="699" alt="image" src="https://user-images.githubusercontent.com/339422/33345845-95bba4fa-d442-11e7-81d6-a9524736ca1c.png">

In this register, the 7-bit `ICI/IT` field is split into a 2 and 5 bit chunk, but it's not ever meaningful to access them as anything but a 7 bit entity. Is there a good way to express this?

If I wrote `ICI_IT MASK(0b11_0000_0000_0_11111), 10 []`, would this do what I want and create a 7-bit field?

Reply:
It would. It would also correctly handle the fields between the two chunks. The only caveat is that when you read the field, it will retain that 9-bit separation between the lower and upper chunk.

Reply:
I think in my dreams there'd be some magic that transparently makes a 7 bit field that's the union of those two, but that's too much magic for a pretty rare corner case :)

To Brad's other question: could you implement that as `NUMBITS(n)`, which internally resolves to `MASK(2**n-1)`?

Reply:
Yeah that would some heavy duty magic.

After taking another look, yes I think it could actually be implemented like that. I think that would be a better interface.

Reply:
I think it should either be `NUMBITS(n) OFFSET(o)` or just `n o`

```
Status [
    TXCOMPLETE  NUMBITS(1) OFFSET(0) [],
    TXINTERRUPT NUMBITS(1) OFFSET(1) [],
    RXCOMPLETE  NUMBITS(1) OFFSET(2) [],
    RXINTERRUPT NUMBITS(1) OFFSET(3) [],
    MODE        NUMBITS(2) OFFSET(4) [
        FullDuplex = 0,
        HalfDuplex = 1,
        Loopback = 2,
        Disabled = 3
    ],
    ERRORCOUNT NUMBITS(2) OFFSET(6) []
]
```

or

```
Status [
    TXCOMPLETE  1 0 [],
    TXINTERRUPT 1 1 [],
    RXCOMPLETE  1 2 [],
    RXINTERRUPT 1 3 [],
    MODE        2 4 [
        FullDuplex = 0,
        HalfDuplex = 1,
        Loopback = 2,
        Disabled = 3
    ],
    ERRORCOUNT 2 6 []
]
```